Allow partial downloads

PrestonFrom opened the pull request:
Add functionality to allow partial downloads of URLs. I'm not sure if you want this in the main project or not, but thought I'd submit it.

  modified:   examples/CMakeLists.txt
  new file:   examples/async_simple_max_bytes.cpp
  modified:   inc/lift/RequestHandle.h
  modified:   inc/lift/RequestPool.h
  modified:   src/RequestHandle.cpp
  modified:   src/RequestPool.cpp

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -79,6 +79,17 @@ class RequestHandle
+     * exceed this amount.)
+     * @param max_bytes     The maximum number of bytes to be written for this request.
+     */
+    auto SetMaxBytes(
```
SetMaxDownloadBytes?

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -183,6 +194,23 @@ class RequestHandle
+     * Set to true if the maximum number of bytes has been written.
+     * @param max_bytes_written     Boolean indicating if the max number of bytes has been written.
+     */
+    auto SetCompletedWritingMaxBytes(bool max_bytes_written) -> void;
```
is this an internal function?

PrestonFrom replied:
Changed to private function (same for getCompletedWritingMaxBytes)

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -198,7 +226,8 @@ class RequestHandle
         CurlPool& curl_pool,
-        OnCompleteHandler on_complete_handler = nullptr
+        OnCompleteHandler on_complete_handler = nullptr,
+        size_t max_bytes = 0
```
max_download_bytes

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -266,6 +300,17 @@ class RequestHandle
     ) -> void; ///< libuv will call this function when the AddRequest() function is called.
+
+    friend auto curl_xfer_info(
+        void *p
```
comment the function as well please
void* p

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -266,6 +300,17 @@ class RequestHandle
+        void *p
+    ) -> int;
+
+    friend auto curl_write_partial_data(
```
comments?

jbaldwin commented on inc/lift/RequestPool.h:
```diff
@@ -68,6 +68,12 @@ class RequestPool
         std::chrono::milliseconds timeout
     ) -> Request;
 
+    auto Produce(
```
does this need a full Produce() override or can it bet a default parameter or just set after the Produce() call is done?  this feels heavy

PrestonFrom replied:
Good point -- default parameter seems like the way to go here.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -65,11 +80,36 @@ auto RequestHandle::init() -> void
+     **/
+    if (m_max_bytes)
+    {
+
```
#if LIBCURL_VERSION_NUM >= 0x072000
CURLOPT_XFERINFOFUNCTION
#else
CURLOPT_PROGRESSFUNCTION
#endif

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -175,6 +215,11 @@ auto RequestHandle::SetVersion(
     }
 }
 
+auto RequestHandle::SetMaxBytes(size_t max_bytes) -> void
```
this seems broken, as it doesn't call init() and thus the CURLOPT_XFERINFOFUNCTION might not be set

PrestonFrom replied:
Totally missed that! Expanded the function to reset all the related variables and set the corresponding write callback and CURLOPT_XFERINFOFUNCTION.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -389,6 +445,18 @@ auto RequestHandle::setCompletionStatus(
+             * has been written, then we intentionally aborted, so let's set
+             * this to success. Otherwise, it's a legitimate error.
+             */
+            if (GetCompletedWritingMaxBytes()) {
```
{ } style, it also seems like these completed writing max bytes functions are internal...

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +553,64 @@ auto curl_write_data(
+     * (Note: This will result in the CURLcode returning as an error.
+     *  This is handled in setCompletionStatus.)
+     **/
+    if (raw_request_ptr->GetCompletedWritingMaxBytes())
```
I would suggest since this is a friend function calling into a private function of the RequestHandle so it can manipulate the max download bytes fields -- they are friends intentionally to keep the API very simple for the end user and by making these public you are leaking internal RequestHandle details through the public API

PrestonFrom replied:
Makes sense! Changed to private functions.

jbaldwin commented on src/RequestPool.cpp:
```diff
@@ -75,6 +75,48 @@ auto RequestPool::Produce(
         return Request(this, std::move(request_handle_ptr));
     }
 }
+auto RequestPool::Produce(
```
OK so it seems like you need to know max_download_bytes when producing since it affects init() and what CURLOPTs get set on the curl handle.  So I would say use an ssize_t and set it to -1 by default, which means "download the entire file", then you do not need another overload on the Produce() function but just an additional defaulted parameter

PrestonFrom replied:
That makes a lot more sense!

jbaldwin commented on examples/CMakeLists.txt:
```diff
@@ -17,6 +17,13 @@ set(ASYNC_SIMPLE_SOURCE_FILES
 add_executable(async_simple ${ASYNC_SIMPLE_SOURCE_FILES})
```
I've added a lot of single comments, please take a look

PrestonFrom replied:
Thank you for the comments! I think I got everything. I wasn't sure if it was better to create a new commit or amend and I ended up creating a new commit. In retrospect, it probably should have just been an amend. If it's difficult to read, please let me know and I'll fix it.

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -227,7 +215,7 @@ class RequestHandle
         CurlPool& curl_pool,
         OnCompleteHandler on_complete_handler = nullptr,
-        size_t max_bytes = 0
+        ssize_t max_download_bytes = -1
```
doc blocks actually look out of date (on_complete_handler and max_download_bytes)

PrestonFrom replied:
Updated!

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -215,9 +216,28 @@ auto RequestHandle::SetVersion(
 }
 
-auto RequestHandle::SetMaxBytes(size_t max_bytes) -> void
+auto RequestHandle::SetMaxDownloadBytes(ssize_t max_download_bytes) -> void
```
So the init() function is duplicating this work, why not just have init() call this function when its setting the max_download_bytes

PrestonFrom replied:
Good point! Added call to init() and removed duplicate code.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -587,14 +610,24 @@ auto curl_write_partial_data(
+    }
+    else
+    {
+        bytes_left_to_write = 0;
```
just initialize the variable to 0 and then you can remove the else block

PrestonFrom replied:
Done!

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -254,8 +242,8 @@ class RequestHandle
-    size_t m_bytes_left_to_write;                ///< Difference between maximum bytes to write and amount written
+    ssize_t m_max_download_bytes;                         ///< Maximum number of bytes to be written.
+    ssize_t m_bytes_left_to_write;               ///< Difference between maximum bytes to write and amount written
     bool m_wrote_max_bytes;                     ///< True if max number of bytes have been written
```
do we really need a bool value for this? its can be calculated by subtracting two fields correct?  I _think_ just having m_max_download_bytes and m_total_bytes_received should be enough to implement this feature with maybe some helper private member functions

PrestonFrom replied:
Good point! Cut down to only m_max_download_bytes and m_bytes_written and add getBytesLeftToWrite.

jbaldwin commented on examples/CMakeLists.txt:
```diff
@@ -17,6 +17,13 @@ set(ASYNC_SIMPLE_SOURCE_FILES
 target_link_libraries(async_simple PRIVATE lifthttp -Wl,-\( ${LIBSSL} ${LIBCRYPTO} ${LIBCURL} ${LIBCARES} -Wl,-\) uv pthread z dl)
 
+### async_simple_max_bytes ###
+set(ASYNC_SIMPLE_SOURCE_FILES
```
this is resetting the ASYNC_SIMPLE_SOURCE_FILES variable, shouldn't it be ASYNC_SIMPLE_MAX_BYTES_SOURCE_FILES?

PrestonFrom replied:
Yes, it should! Fixed!

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -79,6 +79,17 @@ class RequestHandle
+     * @param max_download_bytes     The maximum number of bytes to be written for this request.
+     */
+    auto SetMaxDownloadBytes(
+            ssize_t max_download_bytes
```
nitpicking, but this looks like its tabbed in twice

PrestonFrom replied:
Fixed!

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -191,14 +202,16 @@ class RequestHandle
      * @param request_pool The request pool that generated this handle.
      * @param curl_handle  The CURL* handle for this Request.
      * @param curl_pool    The CurlPool to return the CURL* handle when this request destructs.
+     * @param max_download_bytes    The maximum number of bytes to download, if -1, will download entire file.
```
on_complete_handler? I know it isn't from your code change but it was missing from the doc blocks previously

PrestonFrom replied:
Comment added

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -175,6 +220,14 @@ auto RequestHandle::SetVersion(
+    m_max_download_bytes = max_download_bytes;
+    m_bytes_written = 0;
+
+    init();
```
hmm, this is the opposite of what I suggested, this will do a lot of duplicate work by now resetting a lot of fields on the curl handle that are unrelated to the progress/xfer functions.  I would like to see the reverse as if this is getting set on thousands of handles it _is_ a performance concern

PrestonFrom replied:
Okay, I see where I got confused. This should be correct now.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
+) -> int
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(pointer_to_request);
+
```
return raw_request_ptr->hasDownloadedRequiredBytes() -> int;  This function is duplicated with its sister function

PrestonFrom replied:
Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
+) -> int
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(pointer_to_request);
+
```
return raw_request_ptr->hasDownloadedRequiredBytes() -> int;  This function is duplicated with its sister function

PrestonFrom replied:
Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -401,6 +470,19 @@ auto RequestHandle::onComplete() -> void {
 }
 
+
+auto RequestHandle::getBytesLeftToWrite() -> size_t
```
getDownloadBytesRemaining?  getDownloadBytesLeft?  .. getRemainingDownloadBytes? I know curl uses "write" but its a confusing term in my opinion, is that the number of bytes to write in the REQUEST or "read" in the RESPONSE?

PrestonFrom replied:
Went with getRemainingDownloadBytesOptional, since this now returns an optional as part of bounds checking.

jbaldwin commented on src/RequestPool.cpp:
```diff
@@ -71,6 +74,7 @@ auto RequestPool::Produce(
         request_handle_ptr->SetOnCompleteHandler(on_complete_handler);
         request_handle_ptr->SetUrl(url);
         request_handle_ptr->SetTimeout(timeout);
+        request_handle_ptr->SetMaxDownloadBytes(max_download_bytes);
```
here is your performance penalty

PrestonFrom replied:
Okay, it all makes sense now!

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -389,6 +443,21 @@ auto RequestHandle::setCompletionStatus(
+             * has been written, then we intentionally aborted, so let's set
+             * this to success. Otherwise, it's a legitimate error.
+             */
+            if (getBytesLeftToWrite() == 0)
```
I would suggest this feels like a good use of a ternary since it is only assigning m_status_code

m_status_code = (getRemainingDownloadBytes() == 0)
    ? RequestStatus::SUCCESS
    : RequestStatus::ERROR

PrestonFrom replied:
Updated

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -401,6 +470,19 @@ auto RequestHandle::onComplete() -> void {
+
+auto RequestHandle::getBytesLeftToWrite() -> size_t
+{
+    if (m_max_download_bytes > -1)
```
do you want to do a bounds check and make sure m_bytes_written is not greater than m_max_download_bytes? otherwise you'll underflow (if there is a bug somewhere) and return a very large number

PrestonFrom replied:
Good point! Added bounds checking and it returns an optional -- if false, we have downloaded too many bytes, otherwise it should be true and have the difference.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_xfer_info_old(
```
this isn't that important but maybe just name them after the actual curl functions like all the rest of the curl_ functions? e.g. curl_xfer_info and curl_progress.  Unless if you know which one is the old/new its a bit ambiguous

PrestonFrom replied:
Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -65,11 +84,37 @@ auto RequestHandle::init() -> void
+     * will be preferentially used when both are set.
+     * https://curl.haxx.se/libcurl/c/progressfunc.html
+     **/
+    if (m_max_download_bytes > -1)
```
SetMaxDownloadBytes(max_download_bytes) // done, the max download bytes sets all the appropriate curl_easy_setopt() fields correctly and only once

PrestonFrom replied:
Got it! That makes a lot more sense now.

PrestonFrom commented:
While working on this today, I realized we can abort the call from the write callback -- we don't need to check progress. I've removed that xferinfo/progress callbacks entirely and some of the code they were using. This should cut down on complexity.

Added bounds checking for downloaded bytes vs max bytes using an optional. I'm wondering if this has the potential to cause performance problems though...

jbaldwin commented on src/RequestStatus.cpp:
```diff
@@ -3,15 +3,16 @@
+static const std::string REQUEST_STATUS_TIMEOUT                     = "TIMEOUT";
+static const std::string REQUEST_STATUS_RESPONSE_EMPTY              = "RESPONSE_EMPTY";
+static const std::string REQUEST_STATUS_ERROR                       = "ERROR";
+static const std::string REQUEST_STATUS_DOWNLOADED_TOO_MANY_BYTES   = "DOWNLOADED_TOO_MANY_BYTES";
```
realistically wouldn't you just chop off the extra bytes? i'm a bit confused by this error

PrestonFrom replied:
I was mostly trying to encapsulate an error which might occur when we write more bytes than the set max bytes -- but maybe it would be better to have this indicate a general error in the write callback function?

jbaldwin replied:
It might signify an error in your code, but realistically as an end user the only error you would get is a write error when the remote server fails and doesn't send enough bytes -- I just don't really see it as being very useful since you can always 'chop' off extra bytes if you did happen to save too many

PrestonFrom replied:
Okay, I think I'm on the right page now! Thank you for the comments!

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -401,6 +446,19 @@ auto RequestHandle::onComplete() -> void {
     m_on_complete_handler(std::move(request));
 }
 
+auto RequestHandle::getRemainingDownloadBytesOptional() -> std::optional<size_t>
```
Optional in the function name is redundant, its in the function signature

PrestonFrom replied:
Removed!

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -389,6 +424,16 @@ auto RequestHandle::setCompletionStatus(
+             * Otherwise, we have a bug and downloaded too many bytes.
+             */
+            m_status_code = (getRemainingDownloadBytesOptional())
+                            ? RequestStatus::SUCCESS
```
whitespace

PrestonFrom replied:
I wasn't sure if this is what you had in mind -- I couldn't find another example in the code.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +538,52 @@ auto curl_write_data(
+         * the download to abort. The status will be set to indicate the wrong
+         * number of bytes were downloaded.
+         */
+        ++data_length;
```
this comment seems like the wrong approach, if you HAVE downloaded too many just chop them off and succeed.  realistically you shouldn't have this happen if done correctly but I don't see why you would fail if you did manage to download the requested number of bytes

PrestonFrom replied:
I've updated the code and comment -- I think this is closer to what you had in mind.

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -480,4 +538,52 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_write_partial_data(
```
why not just merge curl_write_data and curl_write_partial_data? i actually can see a nice elegant way where you just keep appending and then IF max download bytes is set you can update the internal member fields and chop/bail out early if you go over

PrestonFrom replied:
Good point! Removed the partial write callback function and integrated the functionality in curl_write_data -- this also meant we didn't need to reset the the callback each time.

jbaldwin commented on inc/lift/RequestHandle.h:
```diff
@@ -79,6 +80,17 @@ class RequestHandle
+     * Sets the maximum number of bytes of data to write.
+     * (The number of bytes downloaded may be greater than the set amount,
+     * but the number of bytes written for the response's data will not
+     * exceed this amount.)
```
you should mention here how to set it to "infinite" or "the full file"

PrestonFrom replied:
Added! Should I do anything to make it more prominent?

jbaldwin commented on inc/lift/RequestStatus.h:
```diff
@@ -19,7 +19,8 @@ enum class RequestStatus
 
-    ERROR               ///< The request had an error and failed.
+    ERROR,               ///< The request had an error and failed.
+    CURL_DOWNLOAD_ERROR  ///< The request had an error in the CURL write callback.
```
this API uses curl internally but we don't expose any "curl" names through the API, so just make it DOWNLOAD_ERROR?

PrestonFrom replied:
Updated to use DOWNLOAD_ERROR

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -549,66 +501,41 @@ auto curl_write_data(
+         **/
+        if (bytes_left_to_write > -1)
+        {
+            auto unsigned_bytes_left_to_write = static_cast<size_t>(bytes_left_to_write);
```
"u" typically denotes unsigned, e.g. uint32_t, uint64_t, etc

PrestonFrom replied:
Changed to 'ubytes_left_to_write'

jbaldwin commented on src/RequestHandle.cpp:
```diff
@@ -215,9 +179,10 @@ auto RequestHandle::SetVersion(
 {
-    m_max_bytes = max_bytes;
+    m_max_download_bytes = max_download_bytes;
+    m_bytes_written = 0;
```
do you reset this field in Reset()?

PrestonFrom replied:
Ah, missed that!
Reset() now sets m_max_download_bytes to -1 (so it defaults to downloading everything) and m_bytes_written to 0.
